policyeval/broker.go

```diff
@@ -0,0 +1,347 @@
+	pendingEvals map[string]PendingEvaluations
+
+	// enqueuedEvals and enqueuedPolicies are indexes to quickly find if an eval
+	// or a policy is waiting be evaluated.
+	enqueuedEvals    map[string]int
+	enqueuedPolicies map[string]string
+
+	// unack tracks evaluations that have not been ack'd yet.
+	unack map[string]*unackEval
+
+	// waiting tracks Dequeue requests that are blocked waiting for work.
+	waiting map[string]chan struct{}
+}
+
+// unackEval tracks an unacknowledged evaluation along with the Nack timer
+type unackEval struct {
+	Eval      *sdk.ScalingEvaluation
+	Token     string
+	NackTimer *time.Timer
+}
+
+// NewBroker returns a new Broker object.
+func NewBroker(l hclog.Logger, timeout time.Duration, deliveryLimit int) *Broker {
+	return &Broker{
+		logger:           l.Named("broker"),
+		nackTimeout:      timeout,
+		deliveryLimit:    deliveryLimit,
+		pendingEvals:     make(map[string]PendingEvaluations),
+		enqueuedEvals:    make(map[string]int),
+		enqueuedPolicies: make(map[string]string),
+		unack:            make(map[string]*unackEval),
+		waiting:          make(map[string]chan struct{}),
+	}
+}
+
+// Enqueue adds an eval to the broker.
+func (b *Broker) Enqueue(eval *sdk.ScalingEvaluation) {
+	b.l.Lock()
+	defer b.l.Unlock()
+	b.enqueueLocked(eval, "")
+}
+
+func (b *Broker) enqueueLocked(eval *sdk.ScalingEvaluation, token string) {
+	logger := b.logger.With(
+		"eval_id", eval.ID, "policy_id", eval.Policy.ID,
+		"queue", eval.Policy.Type, "token", token)
+
+	logger.Debug("enqueue eval")
+
+	// Check if eval is already enqueued.
+	if _, ok := b.enqueuedEvals[eval.ID]; ok {
+		if token == "" {
+			logger.Debug("eval already enqueued")
+			return
+		}
+	} else {
+		b.enqueuedEvals[eval.ID] = 0
+	}
+
+	queue := eval.Policy.Type
+
+	// Get pending heap for the policy type.
+	pending, ok := b.pendingEvals[queue]
+	if !ok {
+		// Initialize pending heap and waiting channel.
+		pending = PendingEvaluations{}
+		if _, ok := b.waiting[queue]; !ok {
+			b.waiting[queue] = make(chan struct{}, 1)
+		}
+	}
+
+	// Check if an eval for the same policy is already enqueued.
+	pendingEvalID, ok := b.enqueuedPolicies[eval.Policy.ID]
+	if !ok {
+		b.enqueuedPolicies[eval.Policy.ID] = eval.ID
+	} else if pendingEvalID != eval.ID {
+		logger.Debug("policy already enqueued")
+
+		// Policy is waiting to be evaluated, but this could be a newer
+		// evaluation request and the policy could have changed. So update
+		// the pending heap with the new eval.
+		i, pendingEval := pending.GetEvaluation(pendingEvalID)
+		if pendingEval != nil {
+			if eval.CreateTime.After(pendingEval.CreateTime) {
+				logger.Debug("new eval is newer, policy updated")
+				b.enqueuedPolicies[eval.Policy.ID] = eval.ID
+				pending[i] = eval
+				heap.Fix(&pending, i)
+			}
+			return
+		}
+	}
+
+	heap.Push(&pending, eval)
+	b.pendingEvals[queue] = pending
+
+	// Unblock any blocked dequeues.
+	select {
+	case b.waiting[queue] <- struct{}{}:
```

Comment:
are we sure that `b.waiting[queue]` exists in the map? do we need to check-and-create?
i think it's only there because of a race condition where the workers created it before we made our first call to `Enqueue`

Reply:
It's created earlier in this function, when we try to find the heap for this queue from `pendingEvals`. 

But maybe it should be more explicit and without the `pendingEvals` guard?

Reply:
ahh, marvelous. sorry, i put so many comments in here i can't read the code anymore.